MudMenu: Enhance keyboard accessibility and focus management

nccadman19 opened the pull request:
fix(`MudMenu`): enhance keyboard accessibility and focus management for nested menus

- Implements consistent keyboard navigation with Arrow, Enter, Space, Tab, and Escape keys
- Enables full support for multi-level nested menus
- Focus returns to appropriate activator elements on close or navigation exit
- Improves support for both default and custom activator content
- Aligns menu behaviour with [WAI-ARIA Menubar guidelines](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/)
- Refactored menu keyboard handling to reduce cognitive complexity and improve maintainability
- Added comprehensive keyboard navigation tests and helper for menu focus handling
- Moved the `MudElement` out of the div to deal with `tabindex` issues
- Tested locally in `MudBlazor.Docs` to confirm behaviour across nested menu scenarios.
- Access modifiers changed from internal to protected or protected internal where required
- Waiting for assertion added to tests to wait for async state updates.
- Tabindex information added to menu docs for accessibility issues on custom activator

## Description

This PR significantly improves accessibility and keyboard interaction support for `MudMenu` and nested submenus. Previously, menu were not accessible via arrow keys and nested menus lacked reliable focus management and full keyboard navigation, especially when custom activator content or complex menu structures were used.

### 🔧 Key Changes
- **Arrow Key Navigation**:
    - `ArrowDown`/`ArrowUp `cycle through items
    - `ArrowRight` opens a submenu if available, or closes all menus if no submenu exists
    - `ArrowLeft` closes a submenu and returns focus to the parent item
- **Enter/Space Activation**:
    - Pressing `Enter` or `Space` on a menu item activates the item or opens its submenu
    - Initial `Enter` press now correctly activates the first item (`MudMenu` or `MudMenuItem`)
- **Focus Management:**
    - Focus is restored to the correct activator after menu close (e.g. button, icon button, or menu item)
    - Custom activator support using `ElementReference` (`_customActivatorRef`)
- **Improved Accessibility**:
    - Menu behaviour aligned with WAI-ARIA menubar interaction guidelines
    - Logical tab flow now preserved between activators and menus
- **Bug Fixes**:
    - Fixes Enter not activating first item in menu
    - Prevents arrow keys from scrolling the page via `KeyInterceptorOptions`
    - `ArrowRight` now closes all menus if no submenu exists

## How Has This Been Tested?

Tested locally by running the MudBlazor docs site using the edited source code:

- Manual tests in `MudBlazor.Docs` including:
    - Navigating menus with all supported keys
    - Activating items with `Enter`/`Space`
    - Opening/closing submenus correctly
    - Skipping disabled items
    - Returning focus to the correct parent or activator
- Unit test coverage added for `Enter/Space` and key navigation
- Checked logical tab flow and focus behaviour when using:
    - Default `MudButton` activators
    - `MudIconButton`
    - `MudMenuItem`
    - Custom content via `<ActivatorContent>`

## Type of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation (fix or improvement to the website or code docs)

## Checklist

- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [X] I've added relevant tests (manual/visual verification in Docs).

## Related Issues
- Closes #6744 

## Remaining Issues / Clarifications
❗ **Scroll on Menu**:
When arrow down is pressed (when the menu is open) the page behaves as though the menu hasn't yet been opened at all and jumps the page down a fraction before the KeyInterceptor kicks in.
Comments from @versile2:
> You didn't actually, you can't stop browser scroll which this is. You can turn it off and we have mechanisms to do that with LockScroll on the MudOverlay but we have generally allowed the user to scroll the page even with menus open. Without setting lockscroll I don't think you can prevent this. @ScarletKuro is this correct?

nccadman19 committed:
Implemented initial keyboard navigation for MudMenu using arrow keys and Enter. Focus now correctly applies on first open.

nccadman19 committed:
Removed mudmenuref, linking items to their indiviudal references instead
Cleared _menuItems on menu close to ensure clean state on next open
Added Task.Yield in OnAfterRenderAsync to wait for menu items to render before focusing
Added null check for ElementRef.Context in FocusItem to prevent JSInterop error
Tabindex in MudMenuItem updated to deal with disabled menu items

bug: still scrolling the page up/down on arrow movements, tabbing twice on items, tabbing outside of the menu starts at the top of page (default popover)

nccadman19 committed:
Handlemenukeydown updated to handlekeydownasync
KeyInterceptorOptions added to avoid page scroll when up down arrows are used

nccadman19 committed:
feat: Add left/right arrow navigation for nested menus

- Add ArrowRight key support to enter submenus from parent menu items
- Add ArrowLeft key support to exit submenus and return to parent
- Enhance keyboard navigation to work with multi-level nested menu structures
- Update key interceptor to include ArrowLeft, ArrowRight, and Escape keys
- Improve menu item registration to persist between open/close cycles
- Fix focus management issues when reopening menus
- Add Enter/Space key support for opening submenus in addition to regular items

bug: when entering a nest menu it jumps from any other nested menus to the first normal menu. tabbing out of the menu jumps to where the popover is located

nccadman19 committed:
Removed tab key interceptor

nccadman19 committed:
fix(MudMenu): improve tab navigation from menu popover

- Store references to activator components (MudButton, MudIconButton, MudMenuItem)
- Focus activator element when tabbing out of menu to maintain logical tab order
- Handle cases where component references may not be available

nccadman19 committed:
Added _hasFocusedOnce flag to fix initial ArrowDown behaviour in MudMenu navigation

- Introduced _hasFocusedOnce boolean to prevent premature re-focusing of item 0
- Ensured first ArrowDown focuses item 0, and subsequent presses move correctly through the list
- Reset _hasFocusedOnce on menu close to restore expected focus behaviour on reopen

nccadman19 committed:
Added keyboard Enter key support to activate first focused MudMenuItem or MudMenu

- Ensured Enter and Space keys correctly trigger click or submenu open on first focus
- Refactored key handling to support nested MudMenu and MudMenuItem activation
- Improved accessibility and consistency of keyboard navigation in nested menus
- Label added to mudmenuitem within mudmenu

Bugs: arrow down twice on first menu load, not currently working on custom activator content
activator content

nccadman19 committed:
Removed hasfocusedasync, no longer needed to stop jumping down on second view of menu
Formatted cascading value to match rest of page
Final arrow left, arrow right and tabs changed to return to menu overview button as seen here
https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/

Bugs: Custom activator content not tabbing back to button on escape or on tab as the ref would need to be within the activator content or with custom js. If you could let me know how you would like me to deal with this.
Two-way Binding button doesnt return to the button when closed, the tabbing returns to the end of the popover

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Changed from tabs to spaces

nccadman19 committed:
DataTestId added to point MouseOver_Click_ShouldKeepMenuOpen to the corrent item

nccadman19 committed:
As above

nccadman19 committed:
fix: update test selector to handle menu wrapper div

nccadman19 committed:
Refactored menu keyboard handling to reduce cognitive complexity and improve maintainability

- Extracted key-specific logic from HandleKeyDownAsync into separate methods:
  HandleNavigationKeyAsync, HandleActivationKeyAsync, and HandleDismissalKeyAsync
- Improved readability and clarity of control flow while preserving full functionality
- Updated FocusActivatorAsync to catch JSException instead of generic Exception,
  avoiding suppression of unexpected errors and aligning with style guide best practices

nccadman19 committed:
feat(tests): add comprehensive keyboard navigation tests and helper for menu focus handling

Added tests for ArrowDown, ArrowUp, ArrowRight, ArrowLeft, Enter, Tab, and Escape keys
Ensured disabled items are skipped during navigation
Implemented helper GetParentMenuFocusedIndex to simplify parent menu focus assertions
Parent menu updated to internal so it is accessible for the tests
Updated arrow up/down navigation logic to ignore disabled menu items, ensuring focus only moves to enabled items. This improves keyboard accessibility and prevents focus from getting stuck on non-interactive elements.

nccadman19 committed:
Moved the mudelement out of the div to deal with tabindexing issues

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Code cleanup

nccadman19 committed:
Tests added to cover the following part of code:
/// <summary>
+         /// Determines whether the specified menu item or submenu is disabled.
+         /// </summary>
+         private bool GetItemDisabled(object item)
+         {
+             return item switch
+             {
+                 MudMenuItem menuItem => menuItem.GetDisabled(),
+                 MudMenu menu => menu.Disabled,
+                 _ => false
+             };
+         }
+
+         /// <summary>
+         /// Retrieves the <see cref="ElementReference"/> associated with a menu item or submenu
+         /// to allow focus control.
+         /// </summary>
+         private ElementReference GetItemElementRef(object item)
+         {
+             return item switch
+             {
+                 MudMenuItem menuItem => menuItem.ElementRef,
+                 MudMenu menu => menu._menuItemActivator?.ElementRef ?? default,
+                 _ => default
+             };
+         }
+
+         /// <summary>
+         /// Determines whether the specified menu item or submenu activates a nested submenu.
+         /// </summary>
+         private bool GetItemActivatesSubMenu(object item)
+         {
+             return item switch
+             {
+                 MudMenuItem menuItem => _itemToSubmenuMap.ContainsKey(menuItem),
+                 MudMenu menu => true,
+                 _ => false
+             };
+         }
      }

nccadman19 committed:
test: stabilize DataGrid grouping tests with WaitForAssertion

Added WaitForAssertion to ensure popover menu items are rendered before interaction.
This resolves flaky test failures caused by premature clicks before UI is ready.

danielchalmers commented:
This looks great, really high quality! Will try to get things moving and take a deeper look tomorrow

(ignore the label spam, testing something out 😄 )

ScarletKuro commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -752,5 +779,304 @@ public void Dispose()
+        /// <summary>
+        /// Sets focus to the menu item at the specified index, if the index is valid.
+        /// </summary>
+        internal async Task FocusItem(int index)
```
FocusItemAsync

ScarletKuro commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -369,7 +380,7 @@ public MudMenu()
 
         [CascadingParameter]
-        protected MudMenu? ParentMenu { get; set; }
+        internal MudMenu? ParentMenu { get; set; }
```
Why are we changing it to internal? It's turning it into a breaking change in case someone inherits the `MudMenu` and I think we have such people. If you need to access it directly, you should change it to `protected internal`.

nccadman19 replied:
Apologies, was testing another way to make it work, forgot to change it back

ScarletKuro commented on src/MudBlazor/Components/Menu/MudMenuItem.razor.cs:
```diff
@@ -118,7 +118,7 @@ public partial class MudMenuItem : MudComponentBase
         protected string GetHtmlTag() => string.IsNullOrEmpty(Href) ? "div" : "a";
 
-        protected bool GetDisabled() => Disabled || ParentMenu?.Disabled == true;
+        internal bool GetDisabled() => Disabled || ParentMenu?.Disabled == true;
```
`protected internal` otherwise this change will be a breaking change.

nccadman19 committed:
Access modifiers changed from internal to protected or protected internal where required
Ran all tests and MudTextField errored due to no two way binding being present. Changed to Value and TextChanged to collect the text output.

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Waiting for assertion added to tests to wait for async state updates.

ScarletKuro committed:
nits

ScarletKuro committed:
Simplify

ScarletKuro commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -752,5 +779,300 @@ public void Dispose()
+        /// <summary>
+        /// Handles keyboard navigation and interaction logic within the menu for arrow keys.
+        /// </summary>
+        private async Task HandleNavigationKeyAsync(KeyboardEventArgs e)
```
I think this need to be split, too much logic in single method, something like
```C#
private async Task HandleNavigationKeyAsync(KeyboardEventArgs e)
{
    var items = _menuItems.Where(x => !GetItemDisabled(x)).ToList();
    if (items.Count == 0)
        return;

    switch (e.Key)
    {
        case "ArrowDown":
            await MoveFocusAsync(1, items.Count);
            break;

        case "ArrowUp":
            await MoveFocusAsync(-1, items.Count);
            break;

        case "ArrowRight":
            await HandleArrowRightAsync();
            break;

        case "ArrowLeft":
            await HandleArrowLeftAsync();
            break;
    }
}
```
Sonar is complaining about the complexity as well.

nccadman19 committed:
_itemToSubmenuMap updated to use the SubMenus list
Logic broken down as too much was in a single method

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Merge to main

nccadman19 committed:
var items updated to omit activator content and avoid duplication of count. Issue: if user is to add a mudbutton not inside a mudmenuitem it would not count it.
Updated logic to deal with disabled items as per W3C guidance

nccadman19 committed:
Aria disabled label added to mudmenuitems

nccadman19 committed:
Updated test to deal with new disabled key logic

danielchalmers commented on src/MudBlazor.UnitTests/Components/DataGridGroupingTests.cs:
```diff
@@ -321,15 +321,26 @@ public void DataGridGroupingTestBoundAndUnboundScenarios()
+
+            for (int i = 0; i < listItems.Count; i++)
+            {
+                var itemText = listItems[i].Find(".mud-menu-item-text").TextContent;
```
This should be an assertion

danielchalmers commented on src/MudBlazor/Components/Menu/MudMenuItem.razor:
```diff
@@ -2,40 +2,42 @@
+				 Icon="@Icon"
+				 Color="IconColor"
+				 Disabled="GetDisabled()" />
+	}
```
space indentations were replaced with tabs in the razor files

danielchalmers commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -752,5 +778,345 @@ public void Dispose()
+                    ]);
+
+            await KeyInterceptorService.SubscribeAsync(_elementId, options, keyDown: HandleKeyDownAsync);
+        }
```
This can return the task.

```suggestion
        private Task SubscribeToMenuKeyInterceptorAsync()
        {
            // Subscribe key interceptor to prevent default scrolling
            var options = new KeyInterceptorOptions(
                "mud-list",
                [
                    // prevent scrolling page
                    new("ArrowDown", preventDown: "key+none"),
                    new("ArrowUp", preventDown: "key+none"),
                ]);

            return KeyInterceptorService.SubscribeAsync(_elementId, options, keyDown: HandleKeyDownAsync);
        }
```

danielchalmers commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -752,5 +778,345 @@ public void Dispose()
+                    _focusedIndex = (_focusedIndex - 1 + itemCount) % itemCount;
+                }
+            }
+            while (false);
```
Do we need the while loop?

ScarletKuro replied:
Seems useless to me too as it does one tick?

ScarletKuro replied:
Also, I think it could be simplified to
```diff
-if (direction > 0)
-{
-	_focusedIndex = (_focusedIndex + 1) % itemCount;
-}
-else
-{
-	_focusedIndex = (_focusedIndex - 1 + itemCount) % itemCount;
-}
+_focusedIndex = (_focusedIndex + direction + itemCount) % itemCount;
```
It should work the same with +1, -1. It will work differently with +2, -2 but I i think it's is a bonus, as you can customize the step like
```
MoveFocusAsync(5, 10)
```
and fast forward the index by 5, instead of just by 1.

danielchalmers commented on src/MudBlazor/Components/Menu/MudMenu.razor.cs:
```diff
@@ -752,5 +778,345 @@ public void Dispose()
+        /// <summary>
+        /// Moves focus up or down in the menu, skipping disabled items.
+        /// </summary>
+        private async Task MoveFocusAsync(int direction, int itemCount)
```
The `async` can be removed and return `FocusItemAsync(_focusedIndex)` without awaiting

danielchalmers commented on src/MudBlazor.UnitTests/Components/MenuTests.cs:
```diff
@@ -780,5 +785,386 @@ public async Task Menu_PointerEvents_RapidMovement()
+        private int GetParentMenuFocusedIndex(IRenderedComponent<MenuKeydownTest> comp)
+        {
+            return (int)typeof(MudMenu)
+                .GetField("_focusedIndex", BindingFlags.NonPublic | BindingFlags.Instance)
```
We should avoid using reflection to assert private/internal state and prefer testing observable behavior instead through bunit

danielchalmers reviewed (changes requested):
I couldn't find examples of sites focusing the first item when spawned from the cursor. It seems like in that case it should have no visible focus but still allow arrow key navigation.

nccadman19 committed:
Test updated to avoid using reflection to assert private/internal state.
LastInvokedIndex added to the keydowntest to log where the focus is on the menu
Datagrid test updated to an assertion
Refactored MoveFocusAsync and SubscribeToMenuKeyInterceptorAsync to remove async await.

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Is pointer check added to on after render to avoid focusing first item on click

nccadman19 commented:
@danielchalmers Heya, think all the changes requested have been implemented. Changes also made regarding: 

> It seems like in that case it should have no visible focus but still allow arrow key navigation.

nccadman19 committed:
Spacing updated on tests and mudmenuitem

versile2 commented:
Absolutely. I'm in the middle of something now but will either look tonight or first thing tomorrow!

ScarletKuro commented:
> Hey, could you help me with this one part. I cannot for the life of me work out how to get it to not focus on first item `OnClick` but still allow `ArrowDown` to enter the menu whilst also maintaining the hover function and also allowing on `Enter/Space` to enter the first item!

I don't think it's that easy (or even possible) to do in Blazor. I don’t really see a way to determine whether the menu was activated by a mouse click or by pressing Enter/Space.
https://github.com/MudBlazor/MudBlazor/blob/36d90ec01a234ed17f053986462b97d430872967/src/MudBlazor/Components/Menu/MudMenu.razor.cs#L999-L1002

I'd just remove that part and keep the logic consistent regardless of how the menu was opened.
Unless I'm misunderstanding the actual problem or there's something others know about how this could be implemented.

versile2 commented:
> @ScarletKuro
> 
> Hey, could you help me with this one part. I cannot for the life of me work out how to get it to not focus on first item `OnClick` but still allow `ArrowDown` to enter the menu whilst also maintaining the hover function and also allowing on `Enter/Space` to enter the first item! I'm blaming it on Fridays :)
> 
> I have committed spacing updates on `MenuKeyDownTests` and `MudMenuItem.razor` but I won't push until we've sorted the below issue. I have updated VS to use spaces from now on.
> 
> This is as far as I've gotten, but this messes with the hover and doesn't allowed `OnClick` → `ArrowDown` to work.
> 
> ```ruby
> internal async Task FocusItemAsync(int index)
>  {
>      // Focus activator if index is -1
>      if (index == -1)
>      {
>          await FocusActivatorAsync();
>          _isPointerOver = true;
>          return;
>      }
> 
>      if (index >= 0 && index < _menuItems.Count)
>      {
>          var item = _menuItems[index];
> 
>          // Retrieves the cref ElementRef associated with a menu item or submenu to allow focus control.
>          ElementReference elementRef = item switch
>          {
>              MudMenuItem menuItem => menuItem.ElementReference,
>              MudMenu menu => menu._menuItemActivator?.ElementReference ?? default,
>              _ => default
>          };
> 
>          if (elementRef.Context is not null)
>          {
>              await elementRef.FocusAsync();
>          }
>      }
>  }
> ```
> 
> ```ruby
> public async Task CloseMenuAsync()
> {
>     // Discard any pending pointer actions so the menu doesn't re-open or try to close twice.
>     CancelPendingActions();
> 
>     // Recursively close all child menus.
>     foreach (var child in _subMenus.Where(m => m._openState.Value))
>     {
>         await child.CloseMenuAsync();
>     }
> 
>     // Now close this menu itself.
>     _focusedIndex = -1;
>     _isPointerOver = false;
>     _menuItems.Clear();
>     await KeyInterceptorService.UnsubscribeAsync(_elementId);
>     await _openState.SetValueAsync(false);
>     await InvokeAsync(StateHasChanged);
> }
> ```
> 
> ```ruby
>  protected override async Task OnAfterRenderAsync(bool firstRender)
>  {
>      await base.OnAfterRenderAsync(firstRender);
> 
>      if (_openState.Value && _focusedIndex == -1)
>      {
>          await Task.Yield();
>          if (_menuItems.Count > 0)
>          {
>              if (!_isPointerOver)
>              {
>                  _focusedIndex = 0;
>                  await FocusItemAsync(_focusedIndex);
>              }
>              else
>              {
>                  _focusedIndex = -1;
>                  await FocusItemAsync(_focusedIndex);
>              }
>          }
>      }
>  }
> ```

To answer your question I believe the only way I see to do it is to use a flag. Were it me I would likely just have a flag I set when I use arrow keys to open a menu that unsets after focus in onafterrender. And if that flag isn't set don't focus.

As another thought while looking at your code and playing with the examples, if I hit "right arrow" it closes all menus, I believe that behavior is it should close any nested menus but not the menu it's on or it's parent.

nccadman19 committed:
DO THIS FOR RIGHT ARROW AND OTHER DISABLED THING
Updated handlerightarrow to do nothing when no further submenus are to be opened as per w3c guidance:
Note that if the menubar were not present, e.g., the menus were opened from a menubutton, Right Arrow would not do anything when focus is on an item that does not have a submenu.

nccadman19 committed:
Ispointer removed, updating to flags

nccadman19 committed:
Flags added to track keyboard/click interactions

nccadman19 committed:
Remove redundant line

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Menu updated to deal with arrow up/down from menu button click (still moving the page slightly before the key interceptor jumps in)
Menu tests and menukeydown tests updated to reflect menu changes

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Added a null-context check before calling _menuWrapperRef.FocusAsync in
OnAfterRenderAsync.

This prevents InvalidOperationException during bUnit tests while keeping
browser behaviour unchanged.

nccadman19 committed:
Added tests for MudMenu.TrackKeyboardInteraction
Ensured code coverage hits ArrowDown, ArrowUp, and default closed branch

nccadman19 commented:
Updated code to address comments from @ScarletKuro below. Flags added to deal with the focus, only problem I'm still facing is the `MudButton` menu click &rarr; Arrow Up/Down is moving the page slightly before the key interceptor jumps in.

The right arrow has been updated so that it does not close all the menus on arrow right. 

> To answer your question I believe the only way I see to do it is to use a flag. Were it me I would likely just have a flag I set when I use arrow keys to open a menu that unsets after focus in onafterrender. And if that flag isn't set don't focus.
> 
> As another thought while looking at your code and playing with the examples, if I hit "right arrow" it closes all menus, I believe that behavior is it should close any nested menus but not the menu it's on or it's parent.

ScarletKuro commented:
@danielchalmers @versile2 please, re-review the PR.
Also, was the `DebouncedTextFieldRerenderTest.razor` changed by mistake? Seems irrelevant changes for the PR.

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 commented:
@versile2 Heya, I'm going on holiday until Monday but I have started making changes to this based on your comments. I'll get them pushed in Tuesday now so you can review, have a nice weekend :)

versile2 commented:
No worries, just ping me when ready.

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
feat(menu):
Improved keyboard accessibility, activator handling, and robustness across menus
Fixed JS error “no element found for id: …” by guarding UnsubscribeAsync on open state and swallowing JSException when the element was already removed.
Stopped hover-menu “flash/reopen” loop:
Introduced a short close cooldown to prevent immediate re-open while the pointer remains over the activator (initially via Task.Yield()).
Made ActivatorContent fully keyboard-operable for top-level menus:
Rendered the activator as a focusable, button-like element with tabindex="0", role="button", aria-haspopup="true", and aria-expanded.
Added HandleActivatorKeydown so Enter / Space / ArrowDown toggle the menu.
Enabled ActivatorContent for submenu activators:
Rendered custom submenu activators as real MudMenuItems so they register in _menuItems and participate in Arrow navigation.
Added the ability to hide the submenu arrow for custom activators (no arrow glyph when using custom content).
if openstate.value added - This prevents keyboard listeners hanging around (important for WCAG 2.1 “no keyboard trap”), and makes the component robust even in nested/disposed menus.
Tabindex required on items within activator content to allow them to be selectable.
Tests updated to handle the await Task.Yield();

To note: freeze on click to scroll still happening, does the new _activatorRef need to tabindex to be like the menuitems getdisabled, would it be helpful if tabindex=0 to be added to activator content in the documents for users to be aware they need to add it to chips and things if they want it clickable

nccadman19 committed:
Merge branch 'fix/accessible-menu' of https://github.com/nccadman19/MudBlazor into fix/accessible-menu

nccadman19 committed:
Debounced test reverted
Assertion added to datagridtests to deal with task yield on menu
Category added to mudmenuitem

nccadman19 commented:
@versile2 Hey, pushed the changes that were mentioned, copied the main parts below:

- Introduced a short close cooldown to prevent immediate re-open while the pointer remains over the activator (initially via `Task.Yield()`).
- Made `ActivatorContent` fully keyboard-operable for top-level menus:
    - Rendered the activator as a focusable, button-like element with `tabindex="0"`, `role="button"`, `aria-haspopup="true"`, and `aria-expanded`.
    - Added `HandleActivatorKeydown` so `Enter` / `Space` toggle the menu.
    - Rendered custom submenu activators as real `MudMenuItems` so they register in `_menuItems` and participate in `Arrow` navigation.
    - Added the ability to hide the submenu arrow for custom activators (no arrow glyph when using custom content). if `openstate.value` added - This prevents keyboard listeners hanging around (important for WCAG 2.1 “no keyboard trap”), and makes the component robust even in nested/disposed menus.
    - `Tabindex` required on items within activator content to allow them to be selectable.
- Tests updated to handle the await `Task.Yield();`

**To note:** 

- Freeze on click to scroll still happening - run the `UnitTests.Viewer` &rarr; Enter the `MenuKeydownTest` &rarr; Zoom page to 200% &rarr; Click Full Menu Width button &rarr; Press `ArrowDown`. Note that the page scrolls on the first arrow down before it freezes on click to keyboard interaction. 
- **Question:** Does the new _activatorRef need to tabindex to be like the `MenuItem.GetDisabled`
- Would it be helpful if tabindex=0 needing to be added to activator content is in the documentation for users to be aware they need to add it to chips and things if they want it clickable in the menu.

versile2 commented:
> @versile2 Hey, pushed the changes that were mentioned, copied the main parts below:
> 
>     * Introduced a short close cooldown to prevent immediate re-open while the pointer remains over the activator (initially via `Task.Yield()`).
> 
>     * Made `ActivatorContent` fully keyboard-operable for top-level menus:
>       
>       * Rendered the activator as a focusable, button-like element with `tabindex="0"`, `role="button"`, `aria-haspopup="true"`, and `aria-expanded`.
>       * Added `HandleActivatorKeydown` so `Enter` / `Space` toggle the menu.
>       * Rendered custom submenu activators as real `MudMenuItems` so they register in `_menuItems` and participate in `Arrow` navigation.
>       * Added the ability to hide the submenu arrow for custom activators (no arrow glyph when using custom content). if `openstate.value` added - This prevents keyboard listeners hanging around (important for WCAG 2.1 “no keyboard trap”), and makes the component robust even in nested/disposed menus.
>       * `Tabindex` required on items within activator content to allow them to be selectable.
> 
>     * Tests updated to handle the await `Task.Yield();`
> 
> 
> **To note:**
> 
>     * Freeze on click to scroll still happening - run the `UnitTests.Viewer` → Enter the `MenuKeydownTest` → Zoom page to 200% → Click Full Menu Width button → Press `ArrowDown`. Note that the page scrolls on the first arrow down before it freezes on click to keyboard interaction.
> 
>     * **Question:** Does the new _activatorRef need to tabindex to be like the `MenuItem.GetDisabled`
> 
>     * Would it be helpful if tabindex=0 needing to be added to activator content is in the documentation for users to be aware they need to add it to chips and things if they want it clickable in the menu.

Sorry it's been a bit since I've been on! Also don't hesitate to remind me if I'm forgetting something :)

So freeze on click to scroll is the correct behavior, if there is a scrollbar then downarrow scrolls. If that behavior needs to be prevented (and I'm not sure it does since it still activates the first item) it's an option of lockscroll for the popover/mudoverlay portion.

For the TabIndex I feel like we can just set it there on that div like you have since it has an onclick? That would produce the needed behavior and not require the end user to do anything different right? Or do they still need to if not using something with a tabindex already set? I know for V9 we want to simplify and get rid of IActivator so this might be something we kick down the road until then. (soon ish I think)

versile2 commented on src/MudBlazor.UnitTests.Viewer/TestComponents/Menu/MenuEdgeCasesTest.razor:
Can see my previous comment on tabindex, if it's required for non button entities we should include it in Docs.

nccadman19 replied:
So would this need to be added to documentation: If the user adds anything within the activator content that includes `@{ ... }` it will need a `tabindex="0"` for the keyboard navigation to reach it.

versile2 commented on src/MudBlazor/Components/Menu/MudMenu.razor:
```diff
@@ -9,24 +9,52 @@
+                     @onkeydown="HandleActivatorKeydown"
+                     disabled="@Disabled"
+                     class="@ActivatorClassname"
+                     tabindex="0"
```
Relates back to the tabindex questions, does this cover when the user uses a non button like a mudchip? if not does it need to be included?

nccadman19 replied:
Yes, this is fine when it is MudElements being added to the ActivatorContent. 

So this would work fine:

```
<MudMenu>
	<ActivatorContent>
		<MudChip T="string" Size="Size.Small" Label="true" Variant="Variant.Filled" style="font-weight: normal; color: black; background-color: #FED6D3" OnClick="@(() => InvokeItem(41))">Priority 1</MudChip>
	</ActivatorContent>
	<ChildContent>
		<MudMenuItem Label="4.1.1" OnClick="@(() => InvokeItem(411))" />
		<MudMenuItem Label="4.1.2" OnClick="@(() => InvokeItem(412))" />
		<MudMenuItem Label="4.1.3" OnClick="@(() => InvokeItem(413))" />
	</ChildContent>
</MudMenu>
```

But this would need tabindex added to it:

```
<MudMenu Class="ml-4" Dense PositionAtCursor>
    <ActivatorContent>
        @{
            var priority = "1";
            switch (priority)
            {
                case "1":
                    <MudChip T="string" Size="Size.Small" Label="true" Variant="Variant.Filled" style="font-weight: normal; color: black; background-color: #FED6D3" tabindex="0">Priority 1</MudChip>
                    break;
                case "2":
                    <MudChip T="string" Size="Size.Small" Label="true" Variant="Variant.Filled" style="font-weight: normal; color: black; background-color: #FDDFCA" tabindex="0">Priority 2</MudChip>
                    break;
            }
        }
    </ActivatorContent>
    <ChildContent>
        <div style="min-width: 100px">
            <MudMenuItem style="font-weight: normal; color: black; background-color: #FED6D3">Priority 1</MudMenuItem>
        </div>
    </ChildContent>
</MudMenu>
```

versile2 reviewed (approved):
This is more a request for information (specifically in regards to tabindex) than changes so I'm listing it as an approval barring those clarifications.

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

nccadman19 committed:
Accessibility tabindex information added to menu docs

nccadman19 committed:
Merge branch 'dev' into fix/accessible-menu

danielchalmers commented:
First, thank you for all the work on this!

I noticed that arrow keys don't work in left or right click context menus / PositionAtCursor

https://github.com/user-attachments/assets/f4be084f-746c-4833-a7be-5cef08f96a2b

unlike File Explorer, Edge, etc

https://github.com/user-attachments/assets/3a150143-b1f8-44a7-849c-bf5974700caa

And it will scroll on the first arrow key press if activated via the cursor

https://github.com/user-attachments/assets/aa74d539-dc79-440b-ba4b-fc5adc631004

I was also able to get it in a state where every arrow up/down would scroll the page in addition to changing the selection, but could not reproduce

Apologies if this has already been discussed!

danielchalmers reviewed (approved):
I think the benefits far outweigh that small quirk and anyone is free to follow up in the future so let's get this merged. Thank you again for your work - always grateful for more accessibility!
